rfcs/text/design-tokens.md

```diff
@@ -0,0 +1,232 @@
+- **Improve handover process and communication between designers and developers.**
+  
+  Both, developers and designers are going to use the same token name for given property (color, background color, border, padding, margin, transition and so on). In the end, developers don't need to know what the final value will be.
+- **Narrow value set to only needed values.**
+  
+  Design System uses narrow set of values (spacings, colors, typography properties and others). Those are only values that are needed for visual description of the component.
+- **Keep visual consistency across all components of all component libraries.**
+
+There are some restrictions from various sides. Some of them are on designers site (processes, design software or plugin's possibilities), and some of them are on developers site (used libraries, frameworks or processes). Therefore, while introducing Design Tokens, few compromises should be made. This requires openness of developers and designers to changes and suggestions to achieve best results. After all, this is a major change.
+
+# Basic example
+To put it in the simplest words: in practice, the introduction of design tokens in Carbon comes down to withdrawal of using `styled-components`'s `props.theme` object in favour of using css variables
+
+```js
+const OldButton = styled.button`
+  font-size: 14px;
+  padding: ${props => props.theme.space[1]}px ${props => props.theme.space[2]}px;
+  border: none;
+  cursor: pointer;
+  
+  color: ${props => props.theme.colors.whiteMix};
+  background-color: ${props => props.theme.colors.base};
+`;
+```
+
+Design tokens are going to be applied with css variables, like this:
+
+```js
+const NewButton = styled.button`
+  font-size: 14px;
+  padding: var(--sizing100) var(--sizing200);
+  border: none;
+  cursor: pointer;
+  
+  color: var(--colorsContent600);
+  background-color: var(--colorsBase500);
+`;
+```
+
+Such approach will require additional compatibility solutions. You can read more of that in [Detailed design](#detailed-design) section.
+
+## Design Tokens Library
+Design Tokens Library will be available in `@sage/design-tokens` NPM package.
+It is built automatically based on what designers will decide in their Figma Design System Libraries.
+NPM package will contain variable sheets for different platforms and formats. Among others, it will contain several formats of javascript variables (object and consts in CommonJS and ES6 formats along with Typescript definitions).
+
+Since carbon is not the only consumer of the tokens, we decided to publish it in a separate repository and as a separate package. This allows for developers on other platforms to have the same tokens that are available anywhere else.
+
+Design Tokens Library updates shall always be backwards compatible, which means that no tokens should be removed or renamed.
```

Comment:
I think we should also NOT change spacing values. This has caused problems in the past when a layout wraps incorrectly with the new size. What are your thoughts?

Reply:
I see the problem you are talking about and in you may be right that it may be dangerous to change spacing values.

However I consider spacings inside components as a design decisions. And design tokens are taking them off from developers and passes them to Designers. So it is up to them what spacings are required.

And If any changes will require updating some other components, then we need to do it anyway.

Reply:
The spacing of component also affects page layout. 

We've previously discussed having two version of carbon on the page at the same time, in micro frontends, owned by different teams that release at different times.


Assuming a single global style sheet this change, will break many layouts 

```diff
.spacing100 {
-  padding: 4px;
+  padding: 8px;
}
```

I think that `spacing100` should remain `4px` and a new value should be introduced for `8px`.

Alternatively, the stylesheet needs to be scoped to a project. There would be multiple, duplicated styles on the page at any one time. This is closest to what we have today using `uuid` class names from `styled-components`.

```diff
-.spacing100 {
-  padding: 4px;
+#app1 .spacing100 {
+  padding: 8px;
}
```


Please can you document your chosen approach in the RFC.

Reply:
we're going to use css variables instead of classes, so instead of
```
.spacing100 { [...] }
```
it will be:
```
:root {
  --spacing100: 4px;
}

#app {
  --spacing100: 8px
}

.btn.medium.localhchange {
  --spacing: 6px;;
}
```

So there always will have to be certain scope.

Reply:
So you can use variables from global scope, but if you need to, you can introduce your own scope and use different version of DTL for that

Reply:
So in terms of the global scope `--spacing100`, the the global value changes, that could break a lot of layouts. In particular when we have two versions of carbon. One Micro Frontend wants to use the new spacing values, but the other micro frontend layouts would break.

How will we avoid this?

Reply:
As we agreed in conversastion: 
There will be additional Design Tokens Provider implemented for scoped tokens.